bazel: 0.24.0 -> 0.26.0

groodt opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
A few new versions of Bazel have been released. This bumps to 0.26.1 to stay recent.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

matthewbauer commented:
@grahamcofborg build bazel

groodt commented:
Please give it another try @matthewbauer 

I still can't get it building on Darwin, but I enabled sandbox to get past the Skydoc error, so I think it may get a bit further on Linux.

My Darwin error is:
```
In file included from src/main/cpp/blaze_util_darwin.cc:24:
/nix/store/mffyf74zc4h2wic5lwx0kyx7gvwif0c5-Libsystem-osx-10.12.6/include/pthread/spawn.h:65:34: error: unknown type name 'qos_class_t'; did you mean 'au_class_t'?
                                 qos_class_t __qos_class);
                                 ^
/nix/store/mffyf74zc4h2wic5lwx0kyx7gvwif0c5-Libsystem-osx-10.12.6/include/bsm/audit.h:176:19: note: 'au_class_t' declared here
typedef u_int32_t       au_class_t;
                        ^
In file included from src/main/cpp/blaze_util_darwin.cc:24:
/nix/store/mffyf74zc4h2wic5lwx0kyx7gvwif0c5-Libsystem-osx-10.12.6/include/pthread/spawn.h:88:34: error: unknown type name 'qos_class_t'; did you mean 'au_class_t'?
                                 qos_class_t * __restrict __qos_class);
                                 ^
/nix/store/mffyf74zc4h2wic5lwx0kyx7gvwif0c5-Libsystem-osx-10.12.6/include/bsm/audit.h:176:19: note: 'au_class_t' declared here
typedef u_int32_t       au_class_t;
                        ^
2 errors generated.
```

groodt commented:
Build succeeds on Linux, but tests are failing with:
```
src/main/tools/process-wrapper-legacy.cc:58: "execvp(external/remote_java_tools_linux/java_tools/ijar/ijar, ...)": No such file or directory
```

uri-canva commented:
Here's a fix for the Darwin issue:
```diff
diff --git a/pkgs/development/tools/build-managers/bazel/default.nix b/pkgs/development/tools/build-managers/bazel/default.nix
index 3b23ca395e8..3a880541228 100644
--- a/pkgs/development/tools/build-managers/bazel/default.nix
+++ b/pkgs/development/tools/build-managers/bazel/default.nix
@@ -160,6 +160,9 @@ stdenv.mkDerivation rec {
         src/tools/xcode/realpath/BUILD \
         src/tools/xcode/stdredirect/BUILD \
         tools/osx/BUILD
+
+      # nixpkgs's libSystem cannot user pthread headers directly, must import GCD headers instead
+      sed -i -e "/#include <pthread\/spawn.h>/i #include <dispatch/dispatch.h>" src/main/cpp/blaze_util_darwin.cc

       # clang installed from Xcode has a compatibility wrapper that forwards
       # invocations of gcc to clang, but vanilla clang doesn't
```

groodt commented:
Legend! Thanks @uri-canva I'll kick off a local test now. I think I'll have to fetch the `java_tools_javac10_darwin-v3.1.zip` as well.

I think I understand what is happening during the `installCheckPhase`. I've got it working if I build on Linux without Nix. Bazel downloads `java_tools_javac10_linux-v3.1.zip` when building the interface jars for junit4 used in the tests.

This is failing under Nix. I'm not 100% clear on how or if this should be fixed. I wonder if the whole approach to these tests may fail, given that Bazel can fetch rules remotely.

Lassulus commented:
@GrahamcOfBorg build bazel

groodt commented:
Success on Darwin, failure on Linux. Must be a first! 😂 

I'm looking into the Linux errors. I didn't have sandbox enabled locally, so I'll do that for Linux which will hopefully help track down errors.

matthewbauer commented:
@uri-canva Thanks! Opened this issue to track what's going on: https://github.com/NixOS/nixpkgs/issues/62555

groodt commented:
@matthewbauer Any chance you can kick off another build. Linux only. I'm unable to run sandboxed builds in a Docker container at the moment. I've fixed the previous error with rules_sass in https://github.com/NixOS/nixpkgs/pull/62336/commits/f1a511a86a02295e2385da96180e3481d68b26d5

uri-canva commented:
I suspect the Bazel tests fail on Linux because of the network sandboxing. The way the building of the tests should be done to account for that is the same way `buildBazelPackage` does it: first run `bazel fetch` as a fixed output derivation with networking enabled, then run `bazel test` with networking disabled.

We could rewrite the whole derivation to split it in two: first build the bootstrap bazel, then use boostrap bazel with `buildBazelPackage` to build and test Bazel. That also mirrors what `compile.sh` does.

It's a lot of work though, we might want to skip testing java on Linux in the meantime.

groodt commented:
@Profpatsch This is now ready for review. Please kick off a build to confirm.

I expect it to pass on Darwin. Fail on Linux. I'm able to reproduce the Linux failure on my local VM.

groodt commented:
> We could rewrite the whole derivation to split it in two: first build the bootstrap bazel, then use boostrap bazel with `buildBazelPackage` to build and test Bazel. That also mirrors what `compile.sh` does.

Yes, I think given the way Bazel works, this is the only sensible solution. I think I understand what is necessary. Will that second inside `buildBazelPackage` be cached in the Nix store?

uri-canva commented:
Yes, the deps derivation is separate from the build derivation and they get cached separately.

Profpatsch commented:
So we’re skipping 0.25 alltogether? https://github.com/NixOS/nixpkgs/pull/61097

groodt commented:
> So we’re skipping 0.25 alltogether? #61097

The Bazel release cycle is too frequent to do 1:1 Nix releases. I certainly don't have time to do it. The functionality I want is in 0.26. In a week or 2 there will 0.27....

Profpatsch commented:
@GrahamcOfBorg build bazel.tests

Profpatsch commented:
The linux test fails.

groodt commented:
> The linux test fails.

@Profpatsch Yes, thanks, it is expected. We just wanted to confirm if it fails on CI with the same error as we get locally. It does, so that's great, we can hopefully fix it.

The failure is due to the sandbox blocking network access during the tests. Bazel isn't very good to test in this way, as rules can go and download other rules at runtime. What are your thoughts to fixing it?

Profpatsch commented:
> What are your thoughts to fixing it?

By now I’m tempted to disable the checkPhase (`doCheck = false;`) and rely on the manual tests we run afterwards. We can’t keep up with their (shitty) WORKSPACE stuff, they don’t really care about other build systems.

Profpatsch commented:
The problem now is that the fetch arbitrary binaries (`external/remote_java_tools_linux/java_tools/ijar/ijar`) and try to execute them, we must patch those first or add their dependencies to the test environment.

uri-canva commented:
The java tools currently live here: https://github.com/bazelbuild/bazel/tree/master/src/java_tools, but they'll move them to https://github.com/bazelbuild/java_tools eventually. We can build them from source in the nix derivation using the bootstrap bazel that should be possible to build without network access.

uri-canva commented:
Sorry I'm just commenting and not helping much on the PR this time, I don't have enough time to dedicate to it at the moment.

uri-canva commented:
We can also patch out some of the remote files as long as we delete the references to them: I cannot believe the `rules_sass` and `rules_nodejs` are needed for building Bazel itself, they're probably used to render the documentation.

groodt committed:
bazel: 0.24.0 -> 0.26.0

groodt commented:
@Profpatsch 

> By now I’m tempted to disable the checkPhase (`doCheck = false;`) and rely on the manual tests we run afterwards.

I agree that this is the best course of action for now. PR updated to reflect this. Passing locally on my Linux VM.

I'll try to find some time to submit an additional PR to change the approach of the Bazel derivation as @uri-canva is suggesting.

kalbasit commented:
@GrahamcOfBorg build bazel

kalbasit commented:
@GrahamcOfBorg build bazel bazel.tests

Profpatsch commented:
Why the *heck* is the builtin rule `py_binary` downloading external libraries.

Profpatsch commented:
I opened an upstream issue https://github.com/bazelbuild/bazel/issues/8575

Profpatsch committed:
bazel: move the python test to `py_binary`

`py_test` tries to download unnecessary dependencies at runtime. We
can just as well run it to check the assertion.

Upstream issue: https://github.com/bazelbuild/bazel/issues/8575

Profpatsch committed:
bazel.tests: prebuild the bazel self-extraction to speed up test

Factor out the common parts of tests & cache the bazel
self-extraction (ugh) to a common store path.

Profpatsch commented:
@GrahamcOfBorg build bazel.tests

Profpatsch committed:
bazel: run the tests on all supported platforms

Profpatsch commented:
@GrahamcOfBorg build bazel.tests

Profpatsch committed:
xe: platforms.linux -> platforms.all

xe is such a trivial package, it should build on every platform that
supports a CC compiler.

Profpatsch commented:
@GrahamcOfBorg build bazel.tests

kalbasit commented on pkgs/development/tools/build-managers/bazel/default.nix:
```diff
@@ -63,34 +93,73 @@ let
+        # yes, this path is kinda magic. Sorry.
+        install_dir=.cache/bazel/_bazel_nixbld
+        mkdir -p $HOME/$install_dir
+        cp -R ${homeDir}/$install_dir/install $HOME/$install_dir
```
the `homeDir` itself already layed out as `$HOME/.cache/...`, why not just `cp -R ${homeDir} $HOME` (obviously rm $HOME before you do)?

Profpatsch replied:
Ah, sorry, I missed the review and force-pushed a few changes with rebases. I had to abstract this out anyway, because bazel *of course* does not use `$HOME` for its extraction on MacOS, but `/var/tmp` …

kalbasit replied:
:man_facepalming: 

Thank you for working on this @Profpatsch!!

kalbasit commented on pkgs/development/tools/build-managers/bazel/default.nix:
```diff
@@ -63,34 +93,73 @@ let
+        # Bazel checks whether the mtime of the install dir files
+        # is >9 years in the future, otherwise it extracts itself again.
+        # see PosixFileMTime::IsUntampered in src/main/cpp/util
+        # What the hell bazel.
```
:rofl:

kalbasit commented on pkgs/development/tools/build-managers/bazel/default.nix:
```diff
@@ -297,7 +365,9 @@ stdenv.mkDerivation rec {
-  doInstallCheck = true;
+  # Temporarily disabling for now. A new approach is needed for this derivation as Bazel
+  # accesses the internet during the tests which fails in a sandbox.
+  doInstallCheck = false;
```
Is this needed if we have `bazel.tests` anyway? Perhaps we can skip this step permanently?

Profpatsch replied:
I think once we have a test for Java & C++ rules, we can get rid of these tests alltogether.

Profpatsch commented:
@GrahamcOfBorg build bazel.tests

kalbasit commented on pkgs/development/tools/build-managers/bazel/default.nix:
```diff
@@ -63,34 +93,81 @@ let
   src = fetchurl {
     url = "https://github.com/bazelbuild/bazel/releases/download/${version}/${name}-dist.zip";
-    sha256 = "11gsc00ghxqkbci8nrflkwq1lcvqawlgkaryj458b24si6bjl7b2";
+    sha256 = "d26dadf62959255d58e523da3448a6222af768fe1224e321b120c1d5bbe4b4f2";
```
Are we standardizing on a different format for the hash?

Profpatsch replied:
Hashes can have multiple encodings, This looks like the first one is baze32 while the second is base16

kalbasit replied:
Nix's standard is base32, at least for now. See https://github.com/NixOS/nix/issues/806#issuecomment-475191490

Profpatsch commented:
@GrahamcOfBorg build bazel.tests

groodt commented:
I very much appreciate the help here @Profpatsch 

`0.26.1` has been released. If you are able to get `0.26.0` working, it might be worth giving one shot to the `0.26.1` since we're in the trenches anyway. I'm happy to help and / or prepare it. I can dig up all the hashes etc. Equally, I'd totally understand if you'd rather not. I get fed up with Bazel shenanigans too.

Profpatsch commented:
Sure, but tomorrow. I think it’s building now (I’m pretty sure caching the bazel extraction for tests was not worth the effort lol).

Profpatsch commented:
@GrahamcOfBorg build bazel.tests

groodt commented:
Great! Seems to have passed. I'll try a quick bump to `0.26.1` sometime today.

groodt commented:
The build works locally for me on Darwin (no sandbox) and a NixOS VM (sandbox = true).

I'll update the title, description and squash the commits.

Profpatsch commented:
Is it okay if we leave the commits as-is? I’d only squash in the fixup commits (`git rebase --autosquash`).

groodt commented:
> Is it okay if we leave the commits as-is? I’d only squash in the fixup commits (`git rebase --autosquash`).

Apologies, I think I'm too late and I've squashed it all down. In some of my other `nixpkgs` contributions, the reviewer was particular that I squashed everything down into a single commit with a particular commit message. Maybe that's not standard across the whole repo or is more nuanced and can be more than a single commit.

Profpatsch commented:
Normally multiple commits for separate changes are not a problem. I’ll force-push the split commits.

Profpatsch commented:
Okay, I rebased to master & autosquashed the fixup commits. Will merge once it’s green again.

Profpatsch commented:
Ah, I’m sorry, I didn’t see the bump to 0.26.1.

kalbasit commented:
I'm putting a quick PR together for 0.26.1.

groodt commented:
Thanks @Profpatsch @kalbasit

timokau commented:
I just noticed that this changed the hash of the fetch phase of `dm-sonnet`. We probably have to be more careful with bazel updates in the future if we're using it in fixed output derivations.

Profpatsch commented:
> I just noticed that this changed the hash of the fetch phase of `dm-sonnet`. We probably have to be more careful with bazel updates in the future if we're using it in fixed output derivations.

Uuuuuuuuuuuuuuugh, I suggest we just don’t.

@timokau But for the while being, can you add a `tests.downstreamPackages = { inherit <list of packages that use `fetchbazel`>; }`?

timokau commented:
Don't do what? Don't be careful seems like a bad idea to me ;) And we're already using it in fixed output derivations, I don't see how we could avoid that when building bazel packages.

Would that actually help? Since they're fixed output, they will just be fetched from the binary cache and therefore pass the tests even though they're broken.

kalbasit commented:
@timokau you can probably override one of the phases of the fetcher to invalidate it's input, and force it to compile. Something like `echo ${version}` should do.

timokau commented:
Interesting idea! I feel like there is a bit of a "just because you can doesn't mean that you should" going on here, but here we go: https://github.com/NixOS/nixpkgs/pull/63645

I have created a separate attribute for the derivation instead of a list, because a list wouldn't get built with `nix-build -A bazel.test`.

Profpatsch commented:
> Don't do what? Don't be careful seems like a bad idea to me ;) And we're already using it in fixed output derivations, I don't see how we could avoid that when building bazel packages.

Don’t ever use fixed output derivations.

timokau commented:
How else should we fetch bazel dependencies?

I'd expect the `fetch` output to become deterministic between releases once bazel is stable. If it still isn't, we can look into how they differ between releases and clean up the fetch result accordingly (similar to what we do with `fetchpatch`).

Profpatsch commented:
> How else should we fetch bazel dependencies?

We do it like with all other (well-behaved) package managers and generate nix expressions for WORKSPACE dependencies. I already wrote a mockup hack for updating bazel proper: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/build-managers/bazel/update-srcDeps.py
But really, bazel devs should pull the stick out of their asses, provide a reasonable interface and become a nice player in the open source world.

uri-canva commented:
Yeah Bazel is really really biased towards monorepos with full vendoring.

I think something good integration-wise can be done with resolved workspaces and provided external repositories, but both mechanisms are very much in their infancy and it's unclear how well supported they are:

https://blog.bazel.build/2018/07/09/bazel-sync-and-resolved-file.html
https://github.com/bazelbuild/bazel/issues/5252

Profpatsch commented:
Ah, yes, that’s the better way tbh. It’s a skylark file, so we should be able to cheat our way by evaling it in a python interpreter.
